build: Clear out stale dependencies when installing requirements

timmc-edx opened the pull request:
This should help prevent some cross-version caching issues in the edx-platform-runner unit tests, which apparently run on a "dirty" environment. (This results in errors between master and older open-releases but could also prevent us from noticing missing deps.)

Calling `make local-requirements` at the end of each `*-requirements` target rather than making it a prerequisite is necessary for using sync, since otherwise the local reqs would be wiped out.

Adding a prerequisite of `pre-requirements` allows us to simplify some workflow calls slightly.

Also fix leading whitespace issue in Makefile.

----

We may wish to backport some or all of these changes to Nutmeg or maybe even Maple due to the caching issues that led to this PR (see discussion in https://github.com/openedx/edx-platform/pull/30759) but the changes will almost certainly not apply cleanly, and will have to be replayed by hand.

timmc-edx committed:
build: Clear out stale dependencies when installing requirements

This should help prevent some cross-version caching issues in the
edx-platform-runner unit tests, which apparently run on a "dirty"
environment. (This results in errors between master and older open-releases
but could also prevent us from noticing missing deps.)

Calling `make local-requirements` at the end of each `*-requirements`
target rather than making it a prerequisite is necessary for using sync,
since otherwise the local reqs would be wiped out.

Adding a prerequisite of `pre-requirements` allows us to simplify some
workflow calls slightly.

The pip lockfile also wasn't being used in the Makefile, so fix that.

Also fix leading whitespace issue in Makefile.

timmc-edx commented on .github/workflows/js-tests.yml:
```diff
@@ -63,7 +63,6 @@ jobs:
 
     - name: Install Required Python Dependencies
       run: |
-        pip install -r requirements/pip.txt
```
All of these removals are just to take advantage of the Make targets now doing this automatically.

timmc-edx commented on Makefile:
```diff
@@ -64,20 +64,24 @@ pull: ## update the Docker image used by "make shell"
+	make local-requirements
 
-test-requirements: local-requirements
-	pip install --exists-action='w' -qr requirements/edx/testing.txt
```
This is the important line, since this target is what's used in `unit-tests.yml` and is where the stale virtualenv issue was coming from.

timmc-edx replied:
Incidentally, I'm not totally sure why `--exists-action=w` is needed here, but if it's not included, then some `-e` dependencies fail to be installed properly when this is run from GH actions.

rgraber replied:
Just checking my understanding: the problem was that when we called local_requirements as a prereq, said requirements were getting overwritten by the next line because that line used pip instead of pip-sync?

timmc-edx replied:
Sort of. Local-requirements was running and installing stuff, and then testing.txt was installing on top of it. There may have been some overwriting going on. But the real problem was that all of this was happening on `edx-platform-runner` instances that were getting _reused_ (I may need to make this clearer in the commit message) and there was nothing removing dependencies from that stale virtualenv when setting up for a new branch! That meant that if this job ran on `master`, a particular plugin was getting installed, and then if it ran on `open-release/maple.master` the build would fail (since that plugin was incompatible with the source code on that branch).

Using `pip-sync` is how we avoid this issue in other situations, and for some reason we'd just never put it in place for CI.

timmc-edx replied:
(Moving the `local-requirements` call to after the `testing.txt` sync is necessary so that the sync doesn't blow away the local-requirements stuff.)

kdmccormick commented on Makefile:
```diff
@@ -64,20 +64,24 @@ pull: ## update the Docker image used by "make shell"
+	pip-sync --pip-args="--exists-action=w" requirements/edx/testing.txt
+	make local-requirements
 
 requirements: pre-requirements ## install development environment requirements
```
Looks like we have some similar but divergent logic between `requirements` and `dev-requirements`. Could you:
* add the `wildcard requirements/edx/private.txt` stuff that is here to `dev-requirements`, and then
* make this a simple alias to `dev-requirements`?

kdmccormick reviewed (commented):
one request, otherwise looks like a great change.

BTW, all these `X-requirements` targets are [brand new](https://github.com/openedx/edx-platform/pull/30890) so don't worry too much about messaging out changes to them.

timmc-edx committed:
fixup! Deduplicate `requirements` target into `dev-requirements`

They were basically doing the same thing, except the newer
`dev-requirements` didn't include private reqs.

kdmccormick commented on Makefile:
```diff
@@ -64,26 +64,28 @@ pull: ## update the Docker image used by "make shell"
+	pip-sync --pip-args="--exists-action=w" requirements/edx/testing.txt
+	make local-requirements
+
+requirements: dev-requirements
```
tiniest of nits: adding this comment back could make it a little clearer that this target installs dev requirements, not base requirements.

```suggestion
requirements: dev-requirements ## install development environment requirements
```

timmc-edx replied:
Oh, yeah. :-)

timmc-edx replied:
and maybe remove it from `dev-requirements`, which is kind of self-explanatory

kdmccormick replied:
👍🏻

kdmccormick reviewed (approved):
One non-blocking nit, otherwise 🚀

timmc-edx committed:
fixup! Move comment back
